### PR TITLE
Add encryption to diskqueue

### PR DESCRIPTION
**Pull request by leehinman**
## What does this PR do?

Adds support for AES 128 CTR encryption to disk queue.

## Why is it important?

Necessary for endpoint running under elastic-agent

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

```
go test -bench=1M -benchtime 1x -count 10 -timeout 600m -benchmem > results.txt

benchstat results.txt
name               time/op
Async1M-16          22.0s ± 1%
EncryptAsync1M-16   20.5s ± 1%
Sync1M-16           25.2s ± 1%
EncryptSync1M-16    27.5s ± 0%

name               alloc/op
Async1M-16         3.43GB ± 0%
EncryptAsync1M-16  3.43GB ± 0%
Sync1M-16          3.42GB ± 0%
EncryptSync1M-16   3.42GB ± 0%

name               allocs/op
Async1M-16          40.6M ± 0%
EncryptAsync1M-16   40.5M ± 0%
Sync1M-16           39.5M ± 0%
EncryptSync1M-16    40.1M ± 0%
```

## Related issues

- Relates elastic-agent-shipper#33
- Relates #31935 
- Relates #31949

### CODE REVIEW

**Comment by cmacknz** [comment]
Benchmarks for this implementation:

```
name               time/op
Async1M-16          22.0s ± 1%
EncryptAsync1M-16   20.5s ± 1%
Sync1M-16           25.2s ± 1%
EncryptSync1M-16    27.5s ± 0%

name               alloc/op
Async1M-16         3.43GB ± 0%
EncryptAsync1M-16  3.43GB ± 0%
Sync1M-16          3.42GB ± 0%
EncryptSync1M-16   3.42GB ± 0%

name               allocs/op
Async1M-16          40.6M ± 0%
EncryptAsync1M-16   40.5M ± 0%
Sync1M-16           39.5M ± 0%
EncryptSync1M-16    40.1M ± 0%
```

Compared to the previous implementation: https://github.com/elastic/beats/pull/31949

```
name           time/op
V1Async1M-16    38.8s ± 2%
V2Async1M-16    37.9s ± 8%
V1Sync1M-16     37.7s ± 1%
V2Sync1M-16     40.3s ± 0%

name           alloc/op
V1Async1M-16   3.47GB ± 0%
V2Async1M-16   3.46GB ± 0%
V1Sync1M-16    3.44GB ± 0%
V2Sync1M-16    3.45GB ± 0%


name           allocs/op
V1Async1M-16    43.4M ± 0%
V2Async1M-16    42.0M ± 0%
V1Sync1M-16     40.5M ± 0%
V2Sync1M-16     41.2M ± 0%
```

This version seems almost twice as fast ad allocates slightly less. What do you think caused this @leehinman? I want to make sure the improvement isn't a mirage :)

**Comment by cmacknz** [comment]
`go test -bench=1M -benchtime 1x -count 10 -timeout 600m -benchmem` Can we make this a mage target or even just a script in the disk queue directory? You've already figured out a reasonable config for running tests and we should save future developers from having to do that themselves.

File: libbeat/publisher/queue/diskqueue/docs/on-disk-structures.md
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/docs/on-disk-structures.md:58]:
```suggestion
flags, which signify options.  The size of options is 8-bits in
```

File: libbeat/publisher/queue/diskqueue/segments.go
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/segments.go:140]: Any reason not to just make this larger for more future proofing? Probably nobody will notice 1-3 extra bytes in each segment file depending on uint16 or uint32 as the larger size.

File: libbeat/publisher/queue/diskqueue/segments.go
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/segments.go:240]:
```suggestion
			"couldn't read header for segment %d: %w", segment.id, err)
```

File: libbeat/publisher/queue/diskqueue/segments_test.go
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/segments_test.go:79]: Is it worth doing one more read to ensure you get `EOF` or whatever the end of segment marker is?

File: libbeat/publisher/queue/diskqueue/segments_test.go
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/segments_test.go:127]: Can you test seeking to invalid locations if there isn't a test for that already?

**Comment by leehinman** [comment]
> This version seems almost twice as fast ad allocates slightly less. What do you think caused this @leehinman? I want to make sure the improvement isn't a mirage :)

So this is because #31935 wasn't merged when the benchmark was run for #39149, that is what accounts for the speed up and drop in allocations.

One thing that is directly comparable between #31935 and this PR is 

"1M_10k-16    21.7s ± 0%" is the same test as "Async1M-16          22.0s ± 1%" in this PR. The times are close enough that I would say the changes in this PR that add the options field, and switches to a segmentReader instead of direct io.File, don't change the speed significantly.  The allocations are the same between the two.

File: libbeat/publisher/queue/diskqueue/encryption.go
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/encryption.go:74]: No, having ciphertext around is acceptable.  We expect Malory to be able to "see" cipher text.  Also I would assume anyone that could view memory of the process could also view the segment file which would have the cipher text as well.

File: libbeat/publisher/queue/diskqueue/segments.go
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/segments.go:140]: I waffled on this.  In the end I picked uint8 because I couldn't think of even 8 options, but we could easily go with uint32, that would match version & count.

File: libbeat/publisher/queue/diskqueue/segments_test.go
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/segments_test.go:77]: good idea.  added

File: libbeat/publisher/queue/diskqueue/segments_test.go
Comment (leehinman) [reply libbeat/publisher/queue/diskqueue/segments_test.go:131]: good idea.  added

**Comment by leehinman** [comment]
> `go test -bench=1M -benchtime 1x -count 10 -timeout 600m -benchmem` Can we make this a mage target or even just a script in the disk queue directory? You've already figured out a reasonable config for running tests and we should save future developers from having to do that themselves.

Short term, I added comment to header of `benchmark_test.go` about how to run.  I'll work with team to see what people want long term.

File: libbeat/publisher/queue/diskqueue/benchmark_test.go
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/benchmark_test.go:18]: 👍

File: libbeat/publisher/queue/diskqueue/segments_test.go
Comment (cmacknz) [inline libbeat/publisher/queue/diskqueue/segments_test.go:166]: Was this supposed to stay?

**Review by cmacknz — APPROVED** [review]
LGTM, wait for @faec to review before merging.

**Review by faec — APPROVED** [review]
looks good, thanks!

**Pull request merged**
